File: seminar-06/C/python/C.py
n = int(input())
sinks = []
is_source = [True] * n

for i in range(n):
    row = list(map(int, input().split()))

    if not any(row):          # нет исходящих рёбер → сток
        sinks.append(i + 1)

    for j in range(n):        # есть входящее ребро → не источник
        if row[j]:
            is_source[j] = False

sources = [i + 1 for i in range(n) if is_source[i]]

print(len(sources))
if len(sources) > 0:
    print(*sources, sep='\n')
print(len(sinks))
if len(sinks) > 0:
    print(*sinks, sep='\n')

File: seminar-06/C/java/C.java
import java.io.*;
import java.util.*;

public class C {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine().trim());

        boolean[] isSource = new boolean[n];
        Arrays.fill(isSource, true);

        List<Integer> sinks = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String[] parts = br.readLine().split(" ");

            boolean hasOut = false; // есть ли исходящие рёбра

            for (int j = 0; j < n; j++) {
                int x = Integer.parseInt(parts[j]);

                if (x == 1) {
                    hasOut = true;       // из i есть ребро
                    isSource[j] = false; // в j входит ребро → не источник
                }
            }

            if (!hasOut) sinks.add(i + 1);
        }

        List<Integer> sources = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (isSource[i]) sources.add(i + 1);
        }

        StringBuilder out = new StringBuilder();

        out.append(sources.size()).append('\n');
        for (int v : sources) out.append(v).append('\n');

        out.append(sinks.size()).append('\n');
        for (int v : sinks) out.append(v).append('\n');

        System.out.print(out);
    }
}